asdf/asdf.py

```diff
@@ -867,6 +868,9 @@ def _open_asdf(
                 self._blocks.read_internal_blocks(fd, past_magic=True, validate_checksums=validate_checksums)
                 self._blocks.read_block_index(fd, self)
 
+            if tree_mapper is not None:
+                tree = tree_mapper(tree)
```

Comment:
This seems like a reasonable place for a `provisional api` warning. I opened a PR with a suggested format: #1290
Another option would be to make the `tree_mapper` argument '_tree_mapper' to discourage it's use in case we need to make changes.